src/jdk.jdwp.agent/share/native/libjdwp/commonRef.c

```diff
@@ -544,6 +558,84 @@ commonRef_unpin(jlong id)
     return error;
 }
 
+/* Prevent garbage collection of object */
+void
+commonRef_pinAll()
+{
+    debugMonitorEnter(gdata->refLock); {
+        gdata->pinAllCount++;
+
+        if (gdata->pinAllCount == 1) {
+            JNIEnv  *env;
+            RefNode *node;
+            RefNode *prev;
+            int     i;
+
+            env = getEnv();
+
+            /*
+             * Walk through the id-based hash table. Detach any nodes
+             * for which the ref has been collected.
+             */
+            for (i = 0; i < gdata->objectsByIDsize; i++) {
+                node = gdata->objectsByID[i];
+                prev = NULL;
+                while (node != NULL) {
+                    jobject strongRef;
+
+                    strongRef = strengthenNode(env, node);
+
+                    /* Has the object been collected? */
+                    if (strongRef == NULL) {
+                        RefNode *freed;
+
+                        /* Detach from the ID list */
+                        if (prev == NULL) {
+                            gdata->objectsByID[i] = node->next;
+                        } else {
+                            prev->next = node->next;
+                        }
+                        freed = node;
+                        node = node->next;
+                        deleteNode(env, freed);
+                    } else {
+                        prev = node;
+                        node = node->next;
+                    }
+                }
+            }
+        }
+    } debugMonitorExit(gdata->refLock);
+}
+
+/* Permit garbage collection of objects */
+void
+commonRef_unpinAll()
+{
+    debugMonitorEnter(gdata->refLock); {
+        gdata->pinAllCount--;
+
+        if (gdata->pinAllCount == 0) {
+            JNIEnv  *env;
+            RefNode *node;
+            int     i;
+
+            env = getEnv();
+
+            for (i = 0; i < gdata->objectsByIDsize; i++) {
+                for (node = gdata->objectsByID[i]; node != NULL; node = node->next) {
+                    jweak weakRef;
+
+                    weakRef = weakenNode(env, node);
+                    if (weakRef == NULL) {
+                        EXIT_ERROR(AGENT_ERROR_NULL_POINTER,"NewWeakGlobalRef");
+                    }
```

Comment:
I'm not so sure I agree that having a fatal error here is the right thing to do. The only other user of `weakenNode()` is `ObjectReference.disableCollection()`. It returns an error to the debugger if `weakenNode()` returns `NULL`. However, I'm not so sure that's a good thing to do here either, since it means the `VM.resume()` will need to fail. Possibly the error should just be ignored, and we live with the ref staying strong.

Reply:
Another options is to save away the weakref in the node when strengthing. This would benefit `ObjectReference.disableCollection()` also, since it would no longer need to deal with a potential OOM. However, I'm not so sure it's actually worth doing. Trying to keep the debug session alive will having allocation errors is probably a fools errand.

Reply:
I was trying to mimic what we already do in `strengthenNode()`, i.e. it's a fatal error if we can't create a JNI ref. Here:

```
        strongRef = JNI_FUNC_PTR(env,NewGlobalRef)(env, node->ref);
        /*
         * NewGlobalRef on a weak ref will return NULL if the weak
         * reference has been collected or if out of memory.
         * It never throws OOM.
         * We need to distinguish those two occurrences.
         */
        if ((strongRef == NULL) && !isSameObject(env, node->ref, NULL)) {
            EXIT_ERROR(AGENT_ERROR_NULL_POINTER,"NewGlobalRef");
        }
```

So it seems appropriate to do the same thing if we fail to create a JNI weak ref. Also, as @plummercj mentioned, if we can't create a JNI ref, continuing the debug session seems rather pointless as we're about to go down anyway (the next allocation in the JVM will be fatal).

Reply:
Okay.